Rider Management

Gegeinthehouse1 opened the pull request:
**📝 Extended Pull Request Description:**
This pull request introduces the initial implementation of the Admin Rider Management System in the dashboard. The following changes and features were added:

🔧 Changes Implemented:
Created Admin\RiderManagementController:

Handles fetching and displaying all registered riders for administrative control.

Created Middleware (EnsureAdmin):

Restricts rider management access to admin users only.

Implemented Blade View (admin/riders.blade.php):

Lists all registered riders.

Provides buttons to Activate, Deactivate, Suspend, and Ban riders.

Includes a link on each rider that navigates to a new page showing all the rides made by the selected user.

🧪 Testing & Verification:
To verify the ride-user relationship and check if the rides appear correctly:

Use Laravel Tinker to test with a user and dummy ride data:

// Retrieve user with ID 2
$user = App\Models\User::find(2);
$user;

// Create a test ride for this user
$ride = App\Models\Rider::create([
    'user_id' => 2,
    'vehicle' => 'ToyotaYaris',
    'pick_up' => 'yes',
    'destination' => 'Sakumono',
    'payment' => 1000,
    'completed' => false,
]);
Then, visit the admin riders page, click the "view rides" link for the user, and confirm that the ride appears.

Gegeinthehouse1 committed:
Rider Management

📝 Extended Pull Request Description:
This pull request introduces the initial implementation of the Admin Rider Management System in the dashboard. The following changes and features were added:

🔧 Changes Implemented:
Created Admin\RiderManagementController:

Handles fetching and displaying all registered riders for administrative control.

Created Middleware (EnsureAdmin):

Restricts rider management access to admin users only.

Implemented Blade View (admin/riders.blade.php):

Lists all registered riders.

Provides buttons to Activate, Deactivate, Suspend, and Ban riders.

Includes a link on each rider that navigates to a new page showing all the rides made by the selected user.

🧪 Testing & Verification:
To verify the ride-user relationship and check if the rides appear correctly:

Use Laravel Tinker to test with a user and dummy ride data:

php
Copy
Edit
// Retrieve user with ID 2
$user = App\Models\User::find(2);
$user;

// Create a test ride for this user
$ride = App\Models\Rider::create([
    'user_id' => 2,
    'vehicle' => 'ToyotaYaris',
    'pick_up' => 'yes',
    'destination' => 'Sakumono',
    'payment' => 1000,
    'completed' => false,
]);
Then, visit the admin riders page, click the "view rides" link for the user, and confirm that the ride appears.

PapaKennethOfficial reviewed (commented):
Great work done Gerard. your code was successfully reviewed, status buttoms worked successfully. One thing which didnt work was when i banned the rider and tried to log in again, it work. But when a rider is banned, he can not have access to the dashboard again. Please review that and work on it.  Good work done by the way. Currently that is the only issue

Gegeinthehouse1 committed:
Pull Request: Admin Rider Management Dashboard

Description

This pull request implements the Admin Dashboard feature to manage riders on the platform, as per the requirements. The dashboard provides a comprehensive view of all registered riders, with functionality to update their status (Activate, Deactivate, Suspend, Ban) and view their ride history. The changes also address a previous issue where rider status was not consistently treated as an integer, which caused functionality issues. The codebase has been updated to ensure status is always an integer, improving reliability.

Changes Made

Backend Updates

RiderManagementController Enhancements (app/Http/Controllers/Admin/RiderManagementController.php):

Modified updateStatus method to cast the status to an integer ($newStatus = (int) $request->status) to ensure consistent data handling.

Added a descriptive status text in the success message using a match expression to provide clear feedback (e.g., "Rider status updated to Active successfully!").

Improved error handling by checking is_rider earlier and returning a specific error message if the user is not a rider.

Ensured that status changes from non-active (status 4) to active are properly validated.

Added logging for errors during status updates to aid debugging.

CheckRiderStatus Middleware (app/Http/Middleware/CheckRiderStatus.php):

Updated to cast the user's status to an integer ($status = (int) $user->status) to handle cases where the status might be stored as a string.

Enhanced error messages using a match expression to provide specific feedback based on the rider's status (e.g., "Your rider account has been banned. Please contact an administrator.").

Added detailed logging to track user authentication, rider status, and middleware decisions.

Rider Model Cleanup (app/Models/Rider.php):

Simplified the isRider, isDriver, and isAdmin methods by removing unnecessary whitespace and improving code readability.

Ensured consistent logic for role and status checks.

Route Definitions (routes/web.php):

Removed the unused confirmStatus route (riders.confirm-status) as it was not implemented in the provided code and replaced by direct status updates.

Added a new route for the rider management dashboard (admin.riders.index) to display all riders.

Ensured the riders.show-rides route is protected by the admin middleware.

Frontend Updates

Rider Management Dashboard View (resources/views/admin/riders/index.blade.php):

Created a new Blade view to display a table of all riders, including their ID, name, email, status, and action buttons.

Implemented buttons for status updates (Activate, Deactivate, Suspend, Ban) with a confirmation prompt to prevent accidental changes.

Added a link to view each rider's ride history, routing to the riders.show-rides endpoint.

Styled the table using Bootstrap for a clean and responsive design.

Ride History View (resources/views/admin/riders/show-rides.blade.php):

Updated the existing ride history view to remove the commented-out <tfoot> section for cleaner code.

Ensured the table displays relevant ride details (Ride ID, Rider, Distance, Date) with proper formatting.

Sign-In Page Improvements (resources/views/auth/login.blade.php):

Added display of session error messages (e.g., for non-active rider status) using an @if (session('error')) block.

Improved form styling and layout for better user experience.

Bug Fixes

Status Integer Casting: Addressed the issue where the rider status was not consistently an integer by explicitly casting it in both the RiderManagementController and CheckRiderStatus middleware. This fixes previous functionality issues where status comparisons failed due to type mismatches.

Non-Rider Error Handling: Added early validation in updateStatus to prevent attempts to update the status of non-rider users, returning a clear error message.

Implementation Details

Rider List Page:

Accessible at /admin/riders (route: admin.riders.index).

Displays a table with columns: Rider ID, Name, Email, Status, Actions.

Status is displayed as a human-readable string (e.g., "Active", "Banned") using a helper function.

Action buttons trigger POST requests to /admin/riders/{rider}/status with the desired status (1: Banned, 2: Suspended, 3: Deactivated, 4: Active).

A "View Rides" link directs to /admin/riders/{rider}/rides.

Ride History Page:

Accessible at /admin/riders/{rider}/rides (route: riders.show-rides).

Displays a table of all rides associated with the rider, including Ride ID, Rider Name, Distance, and Date.

Protected by admin middleware to ensure only authorized users can access it.

Status Update Workflow:

Admins can update a rider's status via buttons on the rider list page.

The updateStatus method validates the input status and ensures the user is a rider before proceeding.

Success and error messages are flashed to the session and displayed on the frontend.

Testing

Manual Testing:

Verified that the rider list page loads all riders correctly and displays their current status.

Tested status updates for each option (Activate, Deactivate, Suspend, Ban) and confirmed that the correct success/error messages are displayed.

Checked the ride history page for a rider and ensured all rides are listed accurately.

Tested the sign-in process for riders with non-active statuses (e.g., Banned) and confirmed they are redirected to the login page with the appropriate error message.

Validated that non-rider users cannot have their status updated via the admin dashboard.

Edge Cases:

Tested with a non-rider user to ensure the error message is displayed.

Verified behavior when attempting to activate a banned rider, ensuring proper validation.

Confirmed that the middleware correctly logs out riders with non-active statuses.

How to Test

Log in as an admin user.

Navigate to /admin/riders to view the rider list.

Click the status buttons (Activate, Deactivate, Suspend, Ban) for a rider and confirm the status updates with the correct success message.

Click "View Rides" for a rider and verify the ride history page loads correctly.

Log in as a rider with a non-active status (e.g., Banned) and confirm you are redirected to the login page with an error message.

Attempt to update the status of a non-rider user and verify the error message is displayed.

PapaKennethOfficial reviewed (changes requested):
your branch has conflicts
